### PR TITLE
Add back CANNODE_NODE_ID for setting a static node ID

### PR DESCRIPTION
**Pull request by AlexKlimaj**
This PR adds back the CANNODE_NODE_ID param for setting a static node ID on cannodes. This will need careful testing as the last time I dove into this, ran into too many edge cases that could prevent a node from booting.

### CODE REVIEW

**Comment by AlexKlimaj** [comment]
@dakejahl @dagar @davids5

File: src/drivers/uavcannode/uavcannode_params.c [hunk 1]
```diff
@@ -85,3 +85,12 @@ PARAM_DEFINE_INT32(CANNODE_PUB_MBD, 0);
...
 PARAM_DEFINE_INT32(CANNODE_PUB_IMU, 0);
+
+/**
+ * UAVCAN CAN node ID (0 for dynamic allocation).
+ *
+ * @min 0
+ * @max 127
+ * @group UAVCAN
+ */
+PARAM_DEFINE_INT32(CANNODE_NODE_ID, 0);
```
Comment (dagar) [inline src/drivers/uavcannode/uavcannode_params.c]: Put it back at the top where it used to be? 

https://github.com/PX4/PX4-Autopilot/blob/1555f2bd2229544c43966ab5f94879c41d8e1e01/src/drivers/uavcannode/uavcannode_params.c#L34-L43

File: src/drivers/uavcannode/UavcanNode.cpp [hunk 2]
```diff
@@ -824,6 +824,20 @@ extern "C" int uavcannode_start(int argc, char *argv[])
 		}
 	}
 
+	// Read the static node ID parameter and use it if no valid node_id from shared memory
+	int32_t cannode_node_id = 0;
+	param_get(param_find("CANNODE_NODE_ID"), &cannode_node_id);
+
+	if (node_id == 0 && cannode_node_id > 0 && cannode_node_id <= 127) {
```
Comment (dagar) [inline src/drivers/uavcannode/UavcanNode.cpp]: At a minimum we should probably print an error if the node ID value isn't valid.

File: src/drivers/uavcannode/UavcanNode.cpp [hunk 3]
```diff
@@ -824,6 +824,27 @@ extern "C" int uavcannode_start(int argc, char *argv[])
...
+	int32_t cannode_node_id = 0;
+	param_get(param_find("CANNODE_NODE_ID"), &cannode_node_id);
+
+	// Check if the static node ID is in range
+	if (cannode_node_id < 0 || cannode_node_id > 127) {
+		PX4_ERR("Invalid static node ID %d, using dynamic allocation", cannode_node_id);
+		cannode_node_id = 0;
+	}
+
+	// Assign the static node ID if no dynamic allocation is used. Do wen't override a valid node ID from the bootloader?
```
Comment (AlexKlimaj) [inline src/drivers/uavcannode/UavcanNode.cpp]: This setup will use the dynamic ID from the bootloader, do we want to override it?

Comment (dakejahl) [reply src/drivers/uavcannode/UavcanNode.cpp]: Yes I think if the user has specified a static node ID via CANNODE_NODE_ID and it does not match the NodeID from the shared memory then we need to override it, print a warning, and update the value in shared memory.

**Review by dakejahl — DISMISSED** [review]
Tested on a couple of our CAN products. Verified setting CANNODE_NODE_ID allows the user to configure the Node ID. If this is later set back to 0 (ID is allocated) the flight controller will continue to allocate that same ID (this is how DNA works). Verified firmware updating continues to work.

Users must take care to avoid configuring two nodes with the same ID.

Need to add docs now.

**Comment by AlexKlimaj** [comment]
In its current state, it will only use the static ID if no ID was allocated on boot. We discussed overriding that if the static ID is set. What do you think?

File: src/drivers/uavcannode/UavcanNode.cpp [hunk 4]
```diff
@@ -824,6 +824,27 @@ extern "C" int uavcannode_start(int argc, char *argv[])
...
+	param_get(param_find("CANNODE_NODE_ID"), &cannode_node_id);
+
+	// Check if the static node ID is in range
+	if (cannode_node_id < 0 || cannode_node_id > 127) {
+		PX4_ERR("Invalid static node ID %ld, using dynamic allocation", cannode_node_id);
+		cannode_node_id = 0;
+	}
+
+	// Assign the static node ID if no dynamic allocation is used. Do wen't override a valid node ID from the bootloader?
+	if (node_id == 0 && cannode_node_id > 0 && cannode_node_id <= 127) {
```
Comment (dagar) [inline src/drivers/uavcannode/UavcanNode.cpp]:
```suggestion
	if (node_id == 0 && cannode_node_id > 0 && cannode_node_id <= uavcan::NodeID::Max) {
```

File: src/drivers/uavcannode/UavcanNode.cpp [hunk 5]
```diff
@@ -824,6 +824,27 @@ extern "C" int uavcannode_start(int argc, char *argv[])
 		}
 	}
 
+	// Read the static node ID parameter and use it if no valid node_id from shared memory
+	int32_t cannode_node_id = 0;
+	param_get(param_find("CANNODE_NODE_ID"), &cannode_node_id);
+
+	// Check if the static node ID is in range
+	if (cannode_node_id < 0 || cannode_node_id > 127) {
```
Comment (dagar) [inline src/drivers/uavcannode/UavcanNode.cpp]:
```suggestion
	if (cannode_node_id < 0 || cannode_node_id > uavcan::NodeID::Max) {
```

File: src/drivers/uavcannode/UavcanNode.cpp [hunk 6]
```diff
@@ -824,6 +824,27 @@ extern "C" int uavcannode_start(int argc, char *argv[])
...
+
+	// Assign the static node ID if no dynamic allocation is used. Do wen't override a valid node ID from the bootloader?
+	if (node_id == 0 && cannode_node_id > 0 && cannode_node_id <= 127) {
+		node_id = cannode_node_id;
+	}
+
+	// Persist the node ID for the bootloader
+	bootloader_app_shared_t shared_write = {};
+	shared_write.node_id = node_id;
+	shared_write.bus_speed = bitrate;
```
Comment (dagar) [inline src/drivers/uavcannode/UavcanNode.cpp]: There's not really any error checking or validation on the bitrate, it might be safer to do this a bit later after the node has actually started spinning successfully. 

For example you could do it here, and only if there was no dynamic node ID allocation. https://github.com/AlexKlimaj/PX4-Autopilot/blob/a4064072c793787aa9aac149c376bbf3c613d8b0/src/drivers/uavcannode/UavcanNode.cpp#L558-L580

**Comment by dakejahl** [comment]
> In its current state, it will only use the static ID if no ID was allocated on boot. We discussed overriding that if the static ID is set. What do you think?

The cannode bootloader always skips waiting for a node ID. Is this what you're referring to? If CANNODE_NODE_ID is set then the cannode app firmware will also skip DNA and boot immediately with the configured ID.

File: src/drivers/uavcannode/UavcanNode.cpp (hunk 6 above)
Comment (dakejahl) [reply src/drivers/uavcannode/UavcanNode.cpp]: Good call! I fixed it and re-tested

**Review by dakejahl — DISMISSED** [review]

File: src/drivers/uavcannode/UavcanNode.cpp [hunk 7]
```diff
@@ -824,6 +833,21 @@ extern "C" int uavcannode_start(int argc, char *argv[])
...
+	param_get(param_find("CANNODE_NODE_ID"), &cannode_node_id);
+
+	// Check if the static node ID is in range
+	if (cannode_node_id < 0 || cannode_node_id > uavcan::NodeID::Max) {
+		PX4_ERR("Invalid static node ID %ld, using dynamic allocation", cannode_node_id);
+		cannode_node_id = 0;
+	}
+
+	// Assign the static node ID if no dynamic allocation is used. Do wen't override a valid node ID from the bootloader?
+	if (node_id == 0 && cannode_node_id > 0 && cannode_node_id <= uavcan::NodeID::Max) {
```
Comment (dakejahl) [inline src/drivers/uavcannode/UavcanNode.cpp]: This logic can actually be simplified by combining with the above range check. The way this works is if a bootloader does not assign an ID (whether statically or through DNA) then the CANNODE_NODE_ID will be used. The default cannode bootloader behavior is to **not** assign an ID and **not** perform allocation. So the only time CANNODE_NODE_ID will not be used is if the bootloader assigns a node ID.

**Review by dakejahl — COMMENTED** [review]
Odd. This works properly with an ARK FPV but doesn't use the configured Node ID with an ARK V6X. When paired with the V6x the cannode will use the previously assigned node ID and not the statically assigned one.

File: src/drivers/uavcannode/UavcanNode.cpp (hunk 6 above)
Comment (dakejahl) [reply src/drivers/uavcannode/UavcanNode.cpp]: this actually changes the behavior. Need to think on it

**Review by dakejahl — CHANGES_REQUESTED** [review]
I'm going to ~open a a fresh PR~ update this PR tomorrow. I think I've got it sorted

**Review by dakejahl — APPROVED** [review]
Alright this is what we want. Previously I was confused about how the bootloader/app shared memory worked and I was debugging why the bootloader was always doing DNA.

This was my confusion
> Shared memory uses CAN filter registers - these are hardware peripheral registers, NOT flash

So with this implementation static node IDs are acheived by the cannode app setting the node ID from the parameter regardless of the ID assigned to the bootloader from DNA. The FC sees the static Node ID and updates the database with the new node ID for this node, and will give out that ID on subesequent boots.  If you move the cannode to another FC it will DNA in the bootloader and receive whatever ID the FC decides, and then the app fw will override that with the static node ID. As a result that first time the node configured with static ID boots and connects to the FC it ends up showing up with one node ID from the bootloader and then the static ID from the app. So the FC "uavcan status" will show 2 nodes with one offline. But from then on out the FC will always DNA the bootloader with that same ID. If the bootloader DNA times out (ARK FPV) then the app always either does DNA or sets a static ID.
